src/ConnectionPool.jl

```diff
@@ -358,115 +383,45 @@ function purge(c::Connection)
     @ensure bytesavailable(c) == 0
 end
 
-"""
-The `pool` is a collection of open `Connection`s.  The `request`
-function calls `getconnection` to retrieve a connection from the
-`pool`.  When the `request` function has written a Request Message
-it calls `closewrite` to signal that the `Connection` can be reused
-for writing (to send the next Request). When the `request` function
-has read the Response Message it calls `closeread` to signal that
-the `Connection` can be reused for reading.
-"""
-const pool = [Vector{Connection}()]
-const poollock = [ReentrantLock()]
-const poolcondition = [Condition()]
-
 """
     closeall()
 
 Close all connections in `pool`.
 """
 function closeall()
-    lock(poollock[Threads.threadid()]) do
-        for c in pool[Threads.threadid()]
-            close(c)
+    for pod in values(POOL.conns)
+        while isready(pod.conns)
+            close(take!(pod.conns))
         end
-        empty!(pool[Threads.threadid()])
+        Threads.atomic_xchg!(pod.numactive, 0)
     end
-    notify(poolcondition[Threads.threadid()])
     return
 end
 
-"""
-    findwritable(type, host, port) -> Vector{Connection}
-
-Find `Connections` in the `pool` that are ready for writing.
-"""
-function findwritable(T::Type,
-                      host::AbstractString,
-                      port::AbstractString,
-                      pipeline_limit::Int,
-                      require_ssl_verification::Bool,
-                      reuse_limit::Int)
-
-    filter(c->(!c.writebusy &&
-               typeof(c.io) == T &&
-               c.host == host &&
-               c.port == port &&
-               c.pipeline_limit == pipeline_limit &&
-               c.require_ssl_verification == require_ssl_verification &&
-               c.writecount < reuse_limit &&
-               c.writecount - c.readcount < pipeline_limit + 1 &&
-               isopen(c.io)), pool[Threads.threadid()])
+struct Pod
+    conns::Channel{Connection}
+    numactive::Threads.Atomic{Int}
```

Comment:
`Base.n_avail(conns)`?

Reply:
Not quite. We don't keep all our connections in `conns`, only ones that are idle/available to re-use. So connections are "owned" by a transaction while there's a request going on, and then `put!` back in `conns` when the transaction finishes. So in order to track max pool size, we need to keep track of the total # of connections that are out there; currently transacting or idle in our `conns`.